hal/halx86/smp/i386/spinup.c

```diff
@@ -0,0 +1,132 @@
+    PVOID SelfPtr;
+    ULONG PageTableRoot;
+    PKPROCESSOR_STATE ProcessorState;
+    KDESCRIPTOR Gdtr;
+} AP_ENTRY_DATA, *PAP_ENTRY_DATA;
+
+typedef struct _AP_SETUP_STACK
+{
+    PVOID ReturnAddr;
+    PVOID KxLoaderBlock;
+} AP_SETUP_STACK, *PAP_SETUP_STACK;
+
+/* FUNCTIONS *****************************************************************/
+
+static
+ULONG
+HalpSetupTemporaryMappings(
+    _In_ PKPROCESSOR_STATE ProcessorState)
+{
+    PMMPDE RootPageTable = ExAllocatePoolZero(NonPagedPool, PAGE_SIZE, TAG_HAL);
+    PMMPDE LowMapPde = Add2Ptr(HalpLowStub, PAGE_SIZE);
+    PMMPTE LowStubPte = MiAddressToPte(HalpLowStub);
+    PHYSICAL_ADDRESS PhysicalAddress;
+    ULONG StartPti;
+
+    /* Copy current mappings */
+    RtlCopyMemory(RootPageTable, MiAddressToPde(NULL), PAGE_SIZE);
+
+    /* Set up low PDE */
+    PhysicalAddress = MmGetPhysicalAddress(LowMapPde);
+    RootPageTable[0].u.Hard.PageFrameNumber = PhysicalAddress.QuadPart >> PAGE_SHIFT;
+    RootPageTable[0].u.Hard.Valid = 1;
+    RootPageTable[0].u.Hard.Write = 1;
+
+    /* Copy low stub PTEs */
+    StartPti = MiAddressToPteOffset(HalpLowStubPhysicalAddress.QuadPart);
+    ASSERT(StartPti + 10 < 1024);
+    for (ULONG i = 0; i < 10; i++)
+    {
+        LowMapPde[StartPti + i] = LowStubPte[i];
+    }
+
+    PhysicalAddress = MmGetPhysicalAddress(RootPageTable);
+    ASSERT(PhysicalAddress.QuadPart < 0x100000000);
+    return (ULONG)PhysicalAddress.QuadPart;
+}
+
+BOOLEAN
+NTAPI
+HalStartNextProcessor(
+    _In_ PLOADER_PARAMETER_BLOCK LoaderBlock,
+    _In_ PKPROCESSOR_STATE ProcessorState)
+{
+    /* Write KeLoaderBlock into Stack */
+    ProcessorState->ContextFrame.Esp = (ULONG)((ULONG_PTR)ProcessorState->ContextFrame.Esp - sizeof(AP_SETUP_STACK));
+    PAP_SETUP_STACK ApStack = (PAP_SETUP_STACK)ProcessorState->ContextFrame.Esp;
+    ApStack->KxLoaderBlock = KeLoaderBlock;
+    ApStack->ReturnAddr = NULL;
+
+    if (HalpStartedProcessorCount == HalpApicInfoTable.ProcessorCount)
+        return FALSE;
+
+    // Initalize the temporary page table
+    // TODO: clean it up after an AP boots successfully
+    ULONG initialCr3 = HalpSetupTemporaryMappings(ProcessorState);
+    if (!initialCr3)
+        return FALSE;
+
+    // Put the bootstrap code into low memory
+    RtlCopyMemory(HalpLowStub, &HalpAPEntry16,  ((ULONG_PTR)&HalpAPEntry16End - (ULONG_PTR)&HalpAPEntry16));
+
+    // Get a pointer to apEntryData
+    PAP_ENTRY_DATA apEntryData = (PVOID)((ULONG_PTR)HalpLowStub + ((ULONG_PTR)&HalpAPEntryData - (ULONG_PTR)&HalpAPEntry16));
+
+    *apEntryData = (AP_ENTRY_DATA){
+        .Jump32Offset = (ULONG)&HalpAPEntry32,
+        .Jump32Segment = (ULONG)ProcessorState->ContextFrame.SegCs,
+        .SelfPtr = (PVOID)apEntryData,
+        .PageTableRoot = initialCr3,
+        .ProcessorState = ProcessorState,
+        .Gdtr = ProcessorState->SpecialRegisters.Gdtr,
+    };
+
+    ApicStartApplicationProcessor(HalpStartedProcessorCount, HalpLowStubPhysicalAddress);
```

Comment:
Are we sure that the CPU will always start no matter what, with this call?

Reply:
SIPI's aren't something one can sanity check very easily. What needs to happen is that the repetitive check for LoaderBlock->Prcb == 0 needs to have a timeout and if it fails it's something weird,

But there's a few reasons I want to leave this as is for now.

1) If we get to this chunk of code the MADT table has reported a higher physical processor count then what increment we are actively starting, thereby if this is writing a processor Number that doesn't exist then the MADT table code is wrong, so I need to know. 

2) The CPU number is valid but we have done something so wrong that some CPU x isn't starting

Honestly would just make it a time out and a DPRINT1.. But not everyone hooks up logs. And as far as we know this shouldn't ever occur.

It would be more useful for me for someone randomly debugging this to get stuck on the ReactOS version debug screen on the SMP builds and come say something instead of it going on and continuing.

tldr:
I'd like to let it just freeze for now while SMP is still so deeply WIP.

Reply:
OK no problem. But then let's keep it as a reminder for later. Maybe you could add an explicit "TODO" comment in the code regarding that (so that we don't forget what might need to be done or improved).